test-workloads/src/bin/guestvm.rs

```diff
@@ -46,6 +46,126 @@ pub fn alloc_error(_layout: Layout) -> ! {
+        "csrrs {vl}, vl, zero",
+        "csrrs {vcsr}, vcsr, zero",
+        "csrrs {vtype}, vtype, zero",
+        "csrrs {vlenb}, vlenb, zero",
+        vlenb = out(reg) vlenb,
+        vl= out(reg) vl,
+        vcsr = out(reg) vcsr,
+        vtype = out(reg) vtype,
+        options(nostack)
+        );
+    }
+    println!("vl    {}", vl);
+    println!("vlenb {}", vlenb);
+    println!("vcsr  {:#x}", vcsr);
+    println!("vtype {:#x}", vtype);
+}
+
+#[cfg(target_feature = "v")]
+pub fn test_vector() {
+    let vec_len: u64 = 8;
+    let vtype: u64 = 0xda;
+    let enable: u64 = 0x200;
+
+    unsafe {
+        // safe because we are only setting a bit in a csr
+        asm!(
+        "csrrs zero, sstatus, {enable}",
+        enable = in(reg) enable,
+        );
+    }
+    println!("Vectors Enabled");
+    print_vector_csrs();
+
+    unsafe {
+        // safe because we are only setting the vector csr's
+        asm!(
+        "vsetvl x0, {vec_len}, {vtype}",
+        vec_len = in(reg) vec_len,
+        vtype = in(reg) vtype,
+        options(nostack),
+        )
+    }
+
+    print_vector_csrs();
+
+    const REG_WIDTH_IN_U64S: usize = 4;
+
+    let mut inbuf = [0_u64; (32 * REG_WIDTH_IN_U64S)];
+    let mut refbuf = [0_u64; (32 * REG_WIDTH_IN_U64S)];
+    for i in 0..inbuf.len() {
+        inbuf[i] = 1 << (i % 53) as u64;
+        refbuf[i] = inbuf[i];
+    }
+
+    let bufp1 = inbuf.as_ptr();
+    let bufp2: *const u64;
+    let bufp3: *const u64;
+    let bufp4: *const u64;
+    unsafe {
+        // safe because we don't go past the length of inbuf
+        bufp2 = bufp1.add(8 * REG_WIDTH_IN_U64S);
+        bufp3 = bufp1.add(16 * REG_WIDTH_IN_U64S);
+        bufp4 = bufp1.add(24 * REG_WIDTH_IN_U64S);
+    }
+    println!("Loading vector registers");
+    unsafe {
+        // safe because the assembly reads into the vector register file
+        asm!(
+        "vl8r.v  v0, ({bufp1})",
+        "vl8r.v  v8, ({bufp2})",
+        "vl8r.v  v16, ({bufp3})",
+        "vl8r.v  v24, ({bufp4})",
+        bufp1 = in(reg) bufp1,
+        bufp2 = in(reg) bufp2,
+        bufp3 = in(reg) bufp3,
+        bufp4 = in(reg) bufp4,
+        options(nostack));
+    }
+
+    print_vector_csrs();
+
+    // Overwrite memory to verify that it changes
+    for i in 0..inbuf.len() {
+        inbuf[i] = 99_u64;
+    }
+
+    println!("reading vector registers");
+    unsafe {
+        // safe because enough memory provided to store entire register file
+        asm! {
+        "vs8r.v  v0, ({bufp1})",
+        "vs8r.v  v8, ({bufp2})",
+        "vs8r.v  v16, ({bufp3})",
+        "vs8r.v  v24, ({bufp4})",
+        bufp1 = in(reg) bufp1,
+        bufp2 = in(reg) bufp2,
+        bufp3 = in(reg) bufp3,
+        bufp4 = in(reg) bufp4,
+        options(nostack)}
+    }
+    print_vector_csrs();
+
+    println!("Verify registers");
+    for i in 0..inbuf.len() {
+        if inbuf[i] != refbuf[i] {
+            println!("error:  {} {} {}", i, refbuf[i], inbuf[i]);
```

Comment:
Should we panic here to make the bug obvious?

Reply:
see above

Reply:
Sure...we we detect a mismatch, dump everything then panic. The advantage that a panic is immediately evident, but prints might get lost in the noise.

Reply:
agreed. I just made sure it displays all of the register contents before panicing